### PR TITLE
[IMP] various: improve the interface of the Activity Plan feature

### PR DESCRIPTION
**Pull request by pyduf**
To improve usability on small device, we add the kanban view for them (only) in the following view:
- mail.activity.plan
- mail.activity.template
- mail.activity.type

Additional improvements:
- button "Schedule activitity" in the activity view: only the text is now clickable, not the whole table cell.
- add the avatar for the default user field in the activity type view and plan template view
- always show res_model in the activity plan form even if it cannot be modified (instead of making it invisible in that case, we make it readonly) and show the company field (if any) on the right of it
- hide department column in plan list view when the default model is not hr.employee (indeed that column has no meaning for other target model)
- improve the action helper for the plan (unfortunately the help is duplicated for each action)
- rephrase "Activity To Generate" to "Activities To Create" in the plan form
- add a placeholder for the plan date in the schedule wizard
- when scheduling a plan, don't display plan without any activities
- add some placeholders in the plan template views
- make column company of activity plan template invisible in the embedded tree view (as the company is the same for all template of a plan but needed for a field domain).
- refine configuration menu about activities in crm, sale and hr
- add a placeholder for plan when scheduling activities + make the field bigger
- reorganise the activity menus in "Settings -> Technical"
- add group by model filter for activity plan
- add a detailed plan tree view in "Settings" as all the activity plans are displayed there (so it is important to see to which model the plan applies)

[IMP] hr{_fleet}, mail: add activity dates in plan summary

When scheduling a plan, to give a better overview of the plan, we add in the plan summary, the computed date of each activity that will be created.

Technical note: we move the plan summary computation from mail_activity_plan to the wizard mail_activity_schedule as it now depends on the plan date to display the actual date of each activity that will be created. We could have kept it in plan and display only the "delay" (ex.: +3 days) but displaying date is more convenient. We also rename it from plan_assignation_summary to plan_summary as it is no longer about who will be assigned to the activity.

[IMP] {test_}mail, hr_contract: allow to configure activity templates deadline

When configuring a plan, we can now set for each activity template a delay to modulate (after/before) the deadline of the created activities when scheduling a plan.

Before, the deadline of the activities created when scheduling an activity plan were determined by the plan date + the delay configured in the related activity type.

We also rewrite slightly the tests to check those dates and we rename plan_date_dateline to plan_date of the activity schedule wizard because that date doesn't represent the deadline but the base date of the plan. Indeed, from that date the activity deadline are computed by adding/subtracting the delay_count (based on delay_unit).

[IMP] mail: improve other activities view

We add the action "Mark as Done" in the form view of other activities (activities for which the user has no access to the underlying record).

We also change slightly the activity search view.

Technical notes:
- we rely on the fact that the "active_ids" are relaying from the list view to the action of the form view to avoid an heavy computation (res_user._get_activity_groups).
- we use target "main" in the action to avoid the activity marked as done (and then deleted) to be included in the breadcrumb.

[IMP] mail: remove note field from plan template tree view

The html note field of the activity template tree view was added to allow quick edition of the activity templates. But with the addition of the delay fields, it is often not long enough for a comfortable edition and also produce non uniform line heights. So we have decided to remove it as there is anyway the possibility to have a default note in the activity type that will be used as default note for the template. Also, the user can edit it through the plan settings where each activity template is edited in a form view which contains the note field.

We also correct in this commit the default value for note, responsible_type and responsible_id (taking the value from the related activity type).

We also update slightly the layout of the template activity form.

[IMP] hr{_recruitment}: add recruitment activity plan

We add recruitment plan to help user to plan activities for applicant.

Technical note: some compute methods had "api.onchange" annotation instead of "api.depends". By correcting this, we had to change the test "test_res_model_compatibility" because it is no longer possible to have a department set on a plan with an incompatble model (because the compute method will set the department to False when changing the model to a model incompatible with departments).

Task-3639909

### CODE REVIEW

**Review by awa-odoo — COMMENTED** [review]
Hello @pyduf 

Left a few comments regarding the first commit.
Will look at the second one on Friday as I'm off tomorrow.

Thanks & cheers!

File: addons/mail/models/mail_activity_plan_template.py [hunk 1]
```diff
@@ -71,6 +97,13 @@ def _compute_responsible_id(self):
             if template.responsible_type != 'other' and template.responsible_id:
                 template.responsible_id = False
 
+    def _calculate_date_deadline(self, plan_date=False):
+        """ Return the deadline of the activity to be created from the plan date. """
+        self.ensure_one()
+        base = plan_date or fields.Date.context_today(self)
+        sign = 1 if self.delay_from == 'after_schedule_date' else -1
```
Comment (awa-odoo) [inline addons/mail/models/mail_activity_plan_template.py:104]: Can't you instead first compute the delta, and then inside a if/else add or subtract it from the base_date?

Would be the same but probably easier to read.

**Comment by pyduf** [comment]
Hello @awa-odoo ,  

Thanks for the review. I have updated the code and answered your comments. 
I will discuss with the PO the 2 mentioned points: whether to use o not html field in tree view and scheduling date vs deadline when scheduling a plan.

File: addons/hr/views/mail_activity_plan_views.xml [hunk 2]
```diff
@@ -52,22 +54,27 @@
             <field name="model">mail.activity.plan</field>
             <field name="inherit_id" ref="mail.mail_activity_plan_view_tree"/>
             <field name="arch" type="xml">
-                <xpath expr="//field[@name='steps_count']" position="after">
-                    <field name="department_id" optional="show"/>
+                <xpath expr="//field[@name='steps_count']" position="before">
+                    <field name="department_id" optional="show"
+                           column_invisible="context.get('default_res_model', False) != 'hr.employee'"/>
```
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/hr/views/mail_activity_plan_views.xml:59]: My point was that if you go through the generic settings menu, you can see the hr.employee plans but you will not have access to that column.

I'm not sure it's very important though, so keep it as it for now.

File: addons/mail/models/mail_activity.py [hunk 3]
```diff
@@ -179,29 +179,23 @@ def _onchange_activity_type_id(self):
...
-            self.date_deadline = self._calculate_date_deadline(self.activity_type_id)
+            self.date_deadline = self._get_date_deadline(self.activity_type_id)
             self.user_id = self.activity_type_id.default_user_id or self.env.user
             if self.activity_type_id.default_note:
                 self.note = self.activity_type_id.default_note
 
     @api.model
-    def _calculate_date_deadline(self, activity_type, force_base_date=None):
-        """ Compute the activity deadline given its type, the force_base_date and the context.
+    def _get_date_deadline(self, activity_type):
```
Comment (awa-odoo) [inline addons/mail/models/mail_activity.py:188]: Sorry, maybe that was unclear, I meant moving this method to the "mail.activity.type" model and make it record based.

So you would call: `activity_type._get_date_deadline()` instead of `env['mail.activity']._get_date_deadline(activity_type)`.

File: addons/mail/views/mail_activity_plan_views.xml [hunk 4]
```diff
@@ -42,24 +42,50 @@
...
                                                options="{'no_quick_create': True}"/>
-                                        <field name="summary"/>
+                                        <field name="summary" placeholder="e.g. Discuss Proposal"/>
                                         <field name="responsible_type"/>
-                                        <field name="responsible_id" readonly="responsible_type != 'other'"/>
-                                        <field name="note"/>
-                                        <field name="company_id" optional="hide"/>
+                                        <field name="responsible_id" readonly="responsible_type != 'other'"
+                                               widget="many2one_avatar_user"/>
+                                        <field name="note" placeholder="e.g. Log a note" optional="hide" width="10rem"/>
```
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/mail/views/mail_activity_plan_views.xml:67]: Well, I don't know the whole context so I will leave it to the PO, but it still feels strange to have to edit HTML in a tree view 😄

**Review by pyduf — COMMENTED** [review]
Hello @awa-odoo , 

As suggested, after having done the last correction, I have greened the task so that it could be tested. We have discussed the naming about activity template deadline and chosen to use the same naming as in event: Interval, Unit, Trigger. To help the user, understanding the plan date, when scheduling a plan, we have added now the date of each activity in the summary (but only when plan date is chosen. I did first all the time but we have changed it later for only hen the plan date is set). That new part is in a separate commit.

File: addons/mail/wizard/mail_activity_schedule.py [hunk 5]
```diff
@@ -150,6 +150,29 @@ def _compute_plan_id(self):
     def _compute_plan_date(self):
         self.plan_date = False
 
+    @api.depends('plan_date', 'plan_id')
+    def _compute_plan_summary(self):
+        template_summaries_list_by_plan = {}
+        for template in self.plan_id.template_ids:
+            template_summary = f": {template.summary}" if template.summary else ''
+            template_summaries_list_by_plan.setdefault(template.plan_id, []).append(
+                (template, template.activity_type_id.name, template_summary))
```
Comment (awa-odoo) [inline addons/mail/wizard/mail_activity_schedule.py:159]: Would say the previous code was a bit more simple.

As we will need to fetch all the "summary" field of all the template_ids anyway, even if we have several times the same templates, the cache should do its work.

File: addons/test_mail/tests/test_mail_activity_plan.py [hunk 6]
```diff
@@ -209,13 +210,14 @@ def test_plan_schedule(self):
                  freeze_time(self.reference_now):
                 # No plan_date specified (-> self.reference_now is used), No responsible specified
                 form = self._instantiate_activity_schedule_wizard(test_records)
-                self.assertFalse(form.plan_assignation_summary)
+                self.assertFalse(form.plan_summary)
                 form.plan_id = self.plan_onboarding
-                self.assertEqual(form.plan_assignation_summary,
-                                 '<ul><li>To-Do: Plan training</li><li>To-Do: Training</li></ul>')
+                self.assertEqual("<ul><li>To-Do: Plan training</li><li>To-Do: Training</li></ul>", form.plan_summary)
+                self.assertEqual(form.plan_summary.count('</li>'), 2)
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_mail_activity_plan.py:216]: Seems unnecessary considering we assert the whole content just above ^^

File: addons/hr/tests/test_mail_activity_plan.py [hunk 7]
```diff
@@ -195,9 +195,10 @@ def test_responsible(self):
             # Happy case
             form = self._instantiate_activity_schedule_wizard(employees)
             form.plan_id = self.plan_onboarding
-            self.assertEqual(form.plan_assignation_summary,
-                             '<ul><li>To-Do: Plan training</li><li>To-Do: Training</li>'
-                             '<li>To-Do: Send feedback to the manager</li></ul>')
+            self.assertIn('Plan training', form.plan_summary)
+            self.assertIn('Training', form.plan_summary)
+            self.assertIn('Send feedback to the manager', form.plan_summary)
+            self.assertEqual(form.plan_summary.count('</li>'), 3)
```
Comment (awa-odoo) [inline addons/hr/tests/test_mail_activity_plan.py:201]: Not sure about counting the tags, can't we assert a little more here above to make sure every line is in its own tag if we want to assert the tags?

File: addons/hr_fleet/tests/test_mail_activity_plan.py [hunk 8]
```diff
@@ -43,8 +43,8 @@ def test_responsible(self):
         # Happy case
         form = self._instantiate_activity_schedule_wizard(employees)
         form.plan_id = self.plan_fleet
-        self.assertEqual(form.plan_assignation_summary,
-                         '<ul><li>To-Do: Car return</li></ul>')
+        self.assertIn('Car return', form.plan_summary)
+        self.assertEqual(form.plan_summary.count('</li>'), 1)
```
Comment (awa-odoo) [inline addons/hr_fleet/tests/test_mail_activity_plan.py:47]: Same here.

File: addons/mail/models/mail_activity.py [hunk 9]
```diff
@@ -465,6 +465,20 @@ def action_done(self):
         parameter, therefore setting context to feedback """
         return self.action_feedback()
 
+    def action_done_other_activities(self):
```
Comment (awa-odoo) [inline addons/mail/models/mail_activity.py:468]: Have to say that one is very cryptic, will leave it for the reviewer who has more context that me 😄

**Review by awa-odoo — COMMENTED** [review]
Hello @pyduf 

Thanks for the followup!
I added a few comments while I was at it, but you can keep it where it is in the pipe of course.

Cheers!

**Review by pyduf — COMMENTED** [review]
Hello @awa-odoo , 

Thanks for the additional comments. I have updated the code and answered your comments.

File: addons/mail/models/mail_activity_plan_template.py [hunk 10]
```diff
@@ -23,6 +25,21 @@ class MailActivityPlanTemplate(models.Model):
         domain="['|', ('res_model', '=', False), '&', ('res_model', '!=', False), ('res_model', '=', parent.res_model)]",
         ondelete='restrict', required=True
     )
+    # Activity type delay fields are ignored in favor of these
+    delay_count = fields.Integer(
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_activity_plan_template.py:29]: Shouldn't we pre-populate with values coming from the activity type ? then people can change it if they want but that way activity type act as default values and can be configured once

File: addons/mail/views/mail_activity_views.xml [hunk 11]
```diff
@@ -200,6 +235,10 @@
         <field name="priority">32</field>
         <field name="arch" type="xml">
             <form string="Log an Activity" create="false" delete="false">
+                <header>
+                    <button string="Mark as Done" class="btn-primary"
+                            name="action_done_other_activities" type="object"/>
```
Comment (tde-banana-odoo) [inline addons/mail/views/mail_activity_views.xml:240]: Maybe (even if verbose, otherwise I thought it was used to mark other activities as done)
```suggestion
                            name="action_done_redirect_to_other" type="object"/>
```

File: addons/mail/wizard/mail_activity_schedule.py [hunk 12]
```diff
@@ -146,8 +147,26 @@ def _compute_plan_id(self):
...
+
+    @api.depends('plan_date', 'plan_id')
+    def _compute_plan_summary(self):
+        for scheduler in self:
+            template_ids = scheduler.plan_id.template_ids
+            if not template_ids:
+                scheduler.plan_summary = False
+                continue
+            summaries = []
+            for template in template_ids:
```
Comment (tde-banana-odoo) [inline addons/mail/wizard/mail_activity_schedule.py:161]: Why not like the original code ?
```suggestion
    def _compute_plan_summary(self):
        self.plan_summary = False
        for scheduler in self.filtered('template_ids'):
            summaries = []
            for template in template_ids:
```

File: addons/mail/wizard/mail_activity_schedule.py [hunk 13]
```diff
@@ -146,8 +147,26 @@ def _compute_plan_id(self):
...
+                summaries.append(Markup('<li>%s</li>') % (
+                    template.activity_type_id.name +
+                    (f": {template.summary}" if template.summary else '') +
+                    # We don't display deadlines when the user doesn't specify a plan_date
+                    (f" ({format_date(self.env, template._get_date_deadline(scheduler.plan_date))})"
+                     if scheduler.plan_date else '')
+                ))
+            scheduler.plan_summary = Markup('<ul>%s</ul>') % Markup().join(summaries)
```
Comment (tde-banana-odoo) [inline addons/mail/wizard/mail_activity_schedule.py:169]: Maybe we could split the string formation (I find combination of "+" and if/else quickly hard to follow, but that's personal taste), like
```suggestion
                summary = template.activity_type_id.name
                if template.summary:
                    summary += f": {template.summary}"
                # We don't display deadlines when the user doesn't specify a plan_date
                if scheduler.plan_date
                    summary += f" {format_date(self.env, template._get_date_deadline(scheduler.plan_date))}"
                summaries.append(Markup('<li>%s</li>') % summary)
            scheduler.plan_summary = Markup('<ul>%s</ul>') % Markup().join(summaries)
```

File: addons/test_mail/tests/test_mail_activity_plan.py [hunk 14]
```diff
@@ -194,40 +208,50 @@ def test_plan_schedule(self):
...
+                self.assertPlanExecution(
+                    self.plan_party, test_records,
+                    expected_deadlines=[(self.reference_now + relativedelta(days=-1)).date(),
+                                        (self.reference_now + relativedelta(days=7)).date()])
 
-                # date_deadline specified, responsible specified
-                force_base_date_deadline = date(2050, 1, 15)
-                force_responsible_id = self.user_admin
+                # plan_date specified, responsible specified
+                plan_date = date(2050, 1, 15)
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_activity_plan.py:230]: That's in a lot of years

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, thanks for the PR ! Quite a big cleaning pass :) added some comments. While adding them you already fixed some of them so not all still apply, but hey, submitting quand même.

Bisous !

File: addons/crm/views/mail_activity_plan_views.xml [hunk 15]
```diff
@@ -4,13 +4,17 @@
...
             <field name="context">{'default_res_model': 'crm.lead'}</field>
             <field name="domain">[('res_model', '=', 'crm.lead')]</field>
             <field name="help" type="html">
                 <p class="o_view_nocontent_smiling_face">
-                    Add a new plan
+                    Create an Activity Plan
+                </p>
+                <p>
+                    Activity plans are used to assign a list of activities in just a few clicks
+                    (e.g. "Onboarding", "Offboarding", ...)
```
[the remark this answers is not in the document]
Comment (pyduf) [reply addons/crm/views/mail_activity_plan_views.xml]: Thanks, indeed, we have now added specific help per model (and sometimes per apps: subscription/sale).

**Review by pyduf — COMMENTED** [review]
Hello @tde-banana-odoo , 

Thanks for the review. I have updated the code and answered your comments.
Note that I haven't yet removed the html field from the tree view as I have posted a message on the task about it.

**Review by pyduf — COMMENTED** [review]
Hello @tde-banana-odoo , 

Thanks for the review and your advice. 
I have updated the code and answered your comment.

Unfortunately, the runbot is still red but the same error happens on the master runbot. So I will retry as soon as the master runbot succeed.

**Comment by pyduf** [comment]
Hello @tde-banana-odoo , 

I still have to review a bit the last commit about the new spec. I will tell you when it is done. Thanks.

**Comment by pyduf** [comment]
> Hello @tde-banana-odoo ,
> 
> I still have to review a bit the last commit about the new spec. I will tell you when it is done. Thanks.

@tde-banana-odoo , I have done the changes I wanted. You can take a look when you have time. Thanks.

**Comment by tde-banana-odoo** [comment]
@robodoo override=ci/security

Reason: usage of Markup excluding user input. Moreover it is mainly a code move from already validated Markup (but better check quand même hein).

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge
